### PR TITLE
Project C++ enums to Intl.js, implement Intl.Collator's "usage" option

### LINKED ISSUE #4007: Follow up #4006: Add caching to other Intl-based toLocaleString functions
Follow up #4006

/cc @jackhorton

### LINKED ISSUE #5648: Chakra does not respect `usage` option for Intl.Collator/String.prototype.toLocaleString
See https://github.com/tc39/ecma402/issues/256. A test262 issue has already been created as well.

### PR DESCRIPTION
**Pull request by jackhorton**
Also, some assorted fixes and cleanup:
- We now correctly set the initSlotCapacity of the Intl native interface object
- Moved around some of the asserts in GetLocaleData to be more defensive. None hit, just noticed they had gaps.
- Added caching to Number.prototype.toLocaleString, which is now ~O(1) when calling with default arguments (locally on a debug build, time spent calling toLocaleString 100 times went from 140ms to 20ms)

Fixes #5648 
Fixes #4007

### CODE REVIEW

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp [hunk 1]
```diff
@@ -1087,11 +1132,13 @@ DEFINE_ISXLOCALEAVAILABLE(PR, uloc)
         if (kind == LocaleDataKind::Collation)
         {
             ScopedUEnumeration collations(ucol_getKeywordValuesForLocale("collation", localeID, false, &status));
-            ICU_ASSERT(status, true);
+            int collationsCount = uenum_count(collations, &status);
+
+            // we expect collationsCount to have at least "standard" and "search" in it
+            ICU_ASSERT(status, collationsCount > 2);
```
Comment (jefgen) [inline lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: I don't if it is guaranteed that all locales will have 2 (unique to the locale) collation types. I think that in some cases the "search" might be aliased to the "default/standard" collation type. 
However, that said, there might be fallback to the root data well, so they might always have at least two, but based on falling back to root.

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp [hunk 2]
```diff
@@ -1836,10 +1883,18 @@ DEFINE_ISXLOCALEAVAILABLE(PR, uloc)
...
             CollatorCaseFirst caseFirst = AssertEnumProperty<CollatorCaseFirst>(state, PropertyIds::caseFirstEnum);
+            JavascriptString *usage = AssertStringProperty(state, PropertyIds::usage);
 
             char localeID[ULOC_FULLNAME_CAPACITY] = { 0 };
             LangtagToLocaleID(langtag, localeID);
 
+            const char16 searchString[] = _u("search");
+            if (usage->BufferEquals(searchString, _countof(searchString) - 1)) // minus the null terminator
+            {
+                uloc_setKeywordValue("collation", "search", localeID, _countof(localeID), &status);
```
Comment (jefgen) [inline lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Hmm, yeah, it is probably better to only set the "search" type, that way the defaults for the locale would be used for the "standard" type.

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp (hunk 1 above)
Comment (jackhorton) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: While debugging, I printed out all of the collation values for a few different locales and they all had at least two; if they don't, then our logic below could try to create a JavascriptArray of -1 elements (only if it has 0 collation values, which maybe is fine?), which would just failfast elsewhere (I hope...)

The Intl spec says that every locale has SearchLocaleData and SortLocaleData and usage _must_ accept both search and sort for all locales, so I think the API was designed with the assumption that everyone would have both. It seems like a rare enough case, at least, that I would be interested in being alerted via a defensive failfast if at some point it doesn't hold.

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp (hunk 2 above)
Comment (jackhorton) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Yeah, the intl spec isn't super clear about whether usage=search or -u-co takes precendence since its working at a different abstraction level with the SortLocaleData and SearchLocaleData; if nothing else, SpiderMonkey refuses to accept -u-co when usage == search

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp (hunk 1 above)
Comment (jefgen) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Thanks Jack. To clarify I mean that not every individual locale (in and of itself) would necessarily have multiple collation types. (Like for example "en" might not anything specific to that locale at all). 
Since there is fallback to the root data, the net effect is that yes I think everybody would have multiple.

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp (hunk 2 above)
Comment (jefgen) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Yeah, it does seem a bit odd.. I was wondering about the following scenario, though perhaps it is an edge case.
For example, what if (somehow) the default user locale was set to something other than the standard type. Then if the usage is "standard", would it make sense to change it for force the standard type? If yes, then how would you get the "default" type? Seems a bit odd.

Comment (jefgen) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Though I guess you have to include the locale in the parameters first, before you can specify the options -- so maybe my scenario doesn't make sense, unless you can somehow pass "und" as the locale, and it picks up the host OS default locale.

Comment (jefgen) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Also good point about what takes precedence as well.
Ex:
`Intl.Collator('de-DE-u-co-phonebk', { usage: 'search' })`

Will this give "search" type collation, or "phonebook" type collation?

Comment (jefgen) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Okay, well if the default locale is limited to be the basename, then this couldn't even happen so I guess my hypothetical scenario is moot.

It's still a bit unclear to me which wins out in the explicit API usage though. 
(` Intl.Collator('de-DE-u-co-phonebk', { usage: 'search' }) `)
You could argue that the explicit option in the 2nd parameter should probably "win" over the locale, as the options being passed in are more specific to the immediate API usage at hand, whereas the locale might be anything and/or coming from anywhere (even externally).


Unrelated to the PR at hand but...

> Intl's default locale is the **basename** of the system's default locale

If that is enforced (or mandated by the spec), then that would unfortunately make it harder (in the future) for anything Intl related to try and pick up on any host OS user settings and/or overrides. For example, imagine if the "default locale" was somehow able to be detected and set to "de-DE-u-co-phonebk" or "en-US-u-hc-h23". This would let the various Intl APIs possibly have insight into the user's preferences in a cross-plat way.

Comment (jackhorton) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: I can ask in the next Intl meeting for clarification, but most questions about Intl for me have usually been resolved by "Do what firefox does" (perhaps @zbraniecki knows more here?)

I believe that the more specific reason is that SearchLocaleData, if considered separate from SortLocaleData, wont have any data about -u-co values (unlike sortlocaledata, which will have data for phonebk, emoji, pinyin, etc). So if we are using SearchLocaleData, the technically-closer-to-the-spec approach would be for our call to `platform.getLocaleData("co")` returning an empty list of supported co values rather than the list returned by `uloc_getKeywordValuesForLocale`. Thats a bit harder to wire up, so we can pretend like we just dont accept any co values if usage == search and achieve the same result.

There was talk of a `navigator.locales` web API (so, provided by browser hosts, not necessarily by the engine itself or nodejs) which would mimic the existing `navigator.languages` but would be a list of Intl.Locale objects with all of those unicode options set. I don't know if there was any progress towards actually introducing a spec for it.

Comment (zbraniecki) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: options bag should take precedence over language tag's extensions keys.

Comment (jefgen) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Thank you for the reply Zibi! :)

Okay, if the options bag takes precedence over the language tag, then if "sort" is set as the usage, should that also be set as well? (ie: Perhaps this should be limited to checking for only "search"?)

Comment (zbraniecki) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Can you rephrase the question pls?

Comment (jackhorton) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: My understanding was that to ICU, `en@collation=standard` was ~equivalent to `en`, so we don't need to do anything special in that case (since standard === sort === the default for both Intl and ICU)

Comment (jefgen) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Right, what about something like this:
`new Intl.Collator('es-u-co-trad', { usage: 'sort' })`

If "sort === standard", and the options bag takes precedence, it sounds like in the above case you should be effectively getting 'es-u-co-standard' instead of 'traditional' -- right?

Comment (jackhorton) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Oh... everyone (chakra/v8/sm) currently agrees that results in locale === es-u-co-trad and usage === sort. But I see your point that if usage is specified it should override -u-co, even if the specified value is sort... I can bring it up on the original tc39 issue that prompted me to implement usage, or bring it up at the meeting.

Comment (zbraniecki) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: > I can bring it up on the original tc39 issue that prompted me to implement usage, or bring it up at the meeting.

Yeah, would be good to discuss it. I guess the issue is that we don't want to alter unicode extensions you passed as we give them back in resolvedOptions. But maybe we should remove them? (those are the locales we matched, and those are the options we matched)?
